src/cmd/services/m3coordinator/downsample/id_pool_types.go

```diff
@@ -186,42 +165,45 @@ func (p *rollupIDProvider) finalize() {
 	}
 }
 
+// Next takes the smallest element across both sets of tags, removing any duplicates between the lists.
 func (p *rollupIDProvider) Next() bool {
-	p.index++
-	return p.index < p.Len()
+	if p.index == len(p.tagPairs) && p.mergeTagsIdx == len(p.mergeTags) {
+		// at the end of both sets
+		return false
+	}
+	if p.index == len(p.tagPairs) {
+		// only merged tags left
+		p.curr = p.mergeTags[p.mergeTagsIdx]
+		p.mergeTagsIdx++
+	} else if p.mergeTagsIdx == len(p.mergeTags) {
+		// only provided tags left
+		p.curr = p.tagPairs[p.index]
+		p.index++
+	} else if bytes.Compare(p.tagPairs[p.index].Name, p.mergeTags[p.mergeTagsIdx].Name) == 0 {
+		// a merge tag exists in the provided tag, advance both to prevent duplicates.
+		p.curr = p.tagPairs[p.index]
+		p.index++
+		p.mergeTagsIdx++
+	} else if bytes.Compare(p.tagPairs[p.index].Name, p.mergeTags[p.mergeTagsIdx].Name) < 0 {
+		// the next provided tag is less
+		p.curr = p.tagPairs[p.index]
+		p.index++
+	} else {
+		// the next merge tag is less
+		p.curr = p.mergeTags[p.mergeTagsIdx]
+		p.mergeTagsIdx++
+	}
+	p.currIdx++
+	return true
 }
 
 func (p *rollupIDProvider) CurrentIndex() int {
-	if p.index >= 0 {
```

Comment:
this was actually broken before, so it's intentionally not backwards compatible.  it would produce 0, 0, 1, 2, ...

the addition to the test proves this.